### DISCUSSION TITLE
FileMesh vs. FileMeshGenerator

### DISCUSSION THREAD

**Discussion by hsheldon** [idaholab/moose#17225] [Q&A General] [answered]
Can someone please explain to me the difference between FileMesh and FileMeshGenerator? 
I have come across a strange situation where initial_from_file_var works fine if I use FileMesh, but does not work if I use FileMeshGenerator. I don't get an error, but the variables are not being initialised when I use FileMeshGenerator to read the mesh.

**Comment by YaqiWang** [comment, marked as answer]
`FileMeshGenerator` is the preferred way for creating meshes. It is part of the new mesh generation system. `FileMesh` could be deprecated in the future if I understand correctly. To make `initial_from_file_var` work with `FileMeshGenerator`, you will need to set this parameter `use_for_exodus_restart` to true in the mesh generator to tell MOOSE that this exodus file can be used for initializing variables later.

**Reply by WilkAndy** [reply]
All this really needs to be documented!

The doc string "True to indicate that the mesh file this generator is reading can be used for restarting variables" seems to say nothing about initializing variables, only "restarting" them.

Ideally, there should be a link to a test/example that shows users how to initialize variables.

Also, this documentation is really hard to understand "users can use parameter exodus_extra_element_integers to load elemental variables for setting extra element integers of the mesh. The names of the extra element integers will be the same as the names of the element variables in the mesh file" .   Another example would be useful.

**Reply by hsheldon** [reply]
Agree the doco could be a bit more clear!

**Reply by hsheldon** [reply]
I've been wondering for a while why the old mesh generation methods haven't been deprecated.

**Comment by YaqiWang** [comment]
> I don't get an error, but the variables are not being initialised when I use FileMeshGenerator to read the mesh.

Forgot replying this part: You should hit an error message saying you are missing `use_for_exodus_restart`. 
For example, in `moose/test/tests/ics/from_exodus_solution/elem_part2.i` if you comment out that parameter, you will hit an error.

**Reply by hsheldon** [reply]
Thanks @YaqiWang, it turns out I last updated MOOSE 2 days before you added that bit to the code! Note to self, must update more often :-) Then I would have got the error message and saved myself some hassle.

**Reply by YaqiWang** [reply]
You are welcome. This code was added by me ;-) Doc is always difficult especially we do not have end users as reviewers.
